File: sample_code/scala/src/test/scala/sample/CrsTransformationTest.scala
package sample

import com.programmerare.crsTransformations.CrsTransformationAdapterLeafFactory
import org.scalatest.matchers.should.Matchers._
import org.scalatest.funsuite.AnyFunSuite

// the import below is from the Java library "crs-transformation-constants"
import com.programmerare.crsConstants.constantsByAreaNameNumber.v10_027.EpsgNumber

// the three imports below are from the core Kotlin library "crs-transformation-adapter-core"
import com.programmerare.crsTransformations.coordinate.CrsCoordinate
import com.programmerare.crsTransformations.coordinate.CrsCoordinateFactory
import com.programmerare.crsTransformations.CrsTransformationResult

// cd sample_code/scala
// sbt test
class CrsTransformationTest extends AnyFunSuite {

  test("Transform coordinates from EPSG 4326 to EPSG 3006 with all CrsTransformationAdapters") {
    val allCrsTransformationAdapters = CrsTransformation.getAllCrsTransformationAdapters()
    val inputCoordinate = CrsTransformation.getInputCoordinate()
    val targetCrsIdentifier = CrsTransformation.getTargetCrsIdentifier()

    // Transformation from EPSG 4326 to EPSG 3006:
    assert(inputCoordinate.getCrsIdentifier.getEpsgNumber === EpsgNumber.WORLD__WGS_84__4326) // 4326 is implicit/default EPSG for latitude/longitude
    assert(targetCrsIdentifier.getEpsgNumber === EpsgNumber.SWEDEN__SWEREF99_TM__3006)

    // The transformations with all 10 crsTransformationAdapters
    
    val expectedTotalNumberOfCrsTransformationAdapters = 10
    
    assertResult(expectedTotalNumberOfCrsTransformationAdapters)(allCrsTransformationAdapters.size)

    val results = CrsTransformation.transform(inputCoordinate, targetCrsIdentifier, allCrsTransformationAdapters)
    assertResult(expectedTotalNumberOfCrsTransformationAdapters)(results.size)
    
    results.foreach { result =>
      assertTransformationResult(result)
    }

    val expectedNumberOfLeafCrsTransformationAdapters = 6
    assertResult(expectedNumberOfLeafCrsTransformationAdapters)(CrsTransformationAdapterLeafFactory.getInstancesOfAllKnownAvailableImplementations().size)
    
  }

  private lazy val expectedOnputCoordinate: CrsCoordinate = CrsCoordinateFactory.yx(6580822.0, 674032.0, EpsgNumber.SWEDEN__SWEREF99_TM__3006)

  private def assertTransformationResult(result: CrsTransformationResult) = {
    val errorMessageWithInformationAboutFailingImplementation = s" Failing implementation: ${result.getCrsTransformationAdapterResultSource.getImplementationType}"

    assertResult(true, errorMessageWithInformationAboutFailingImplementation) {
      result.isSuccess
    }

    val outputCoordinate = result.getOutputCoordinate
    assertResult(EpsgNumber.SWEDEN__SWEREF99_TM__3006, errorMessageWithInformationAboutFailingImplementation) {
      outputCoordinate.getCrsIdentifier.getEpsgNumber
    }

    withClue(errorMessageWithInformationAboutFailingImplementation){
      outputCoordinate.getX shouldEqual expectedOnputCoordinate.getX +- 0.5
    }

    withClue(errorMessageWithInformationAboutFailingImplementation){
      outputCoordinate.getY shouldEqual expectedOnputCoordinate.getY +- 0.5
    }
  }
  
}

File: sample_code/java/src/test/java/java_sample/CrsTransformationTest.java
package java_sample;

import org.testng.annotations.Test;
import org.testng.Assert;

// the package below is from the Java library "crs-transformation-constants"
import com.programmerare.crsConstants.constantsByAreaNameNumber.v10_027.EpsgNumber;

import com.programmerare.crsTransformations.CrsTransformationAdapter;
import com.programmerare.crsTransformations.CrsTransformationResult;
import com.programmerare.crsTransformations.compositeTransformations.CrsTransformationAdapterCompositeFactory;
import com.programmerare.crsTransformations.coordinate.CrsCoordinate;
import com.programmerare.crsTransformations.coordinate.CrsCoordinateFactory;


public class CrsTransformationTest {
    // If you want to see Java tests then there are many JUnit tests of the Kotlin code in the module "crs-transformation-adapter-test" 
    // ( https://github.com/TomasJohansson/crsTransformations/tree/master/crs-transformation-adapter-test/src/test/java )
    
    @Test
    public void coordinateTransformationTest() {
        final int epsgWgs84  = EpsgNumber.WORLD__WGS_84__4326;
        final int epsgSweRef = EpsgNumber.SWEDEN__SWEREF99_TM__3006;

        CrsCoordinate centralStockholmWgs84 = CrsCoordinateFactory.latLon(59.330231, 18.059196, epsgWgs84);

        CrsTransformationAdapter crsTransformationAdapter = CrsTransformationAdapterCompositeFactory.createCrsTransformationMedian();
        // If the Gradle/Maven configuration includes all six adapter implementations, then the 
        // above created 'Composite' implementation will below use all six 'leaf' implementations 
        // and when transforming below, will return a resulting coordinate with a median longitude and a median latitude
        CrsTransformationResult centralStockholmResultSweRef = crsTransformationAdapter.transform(centralStockholmWgs84, epsgSweRef);
        Assert.assertTrue(centralStockholmResultSweRef.isSuccess());
        CrsCoordinate resultCoordinate = centralStockholmResultSweRef.getOutputCoordinate(); // median result
        Assert.assertNotNull(resultCoordinate);
        
        // https://javadoc.io/static/org.testng/testng/7.5/org/testng/Assert.html#assertEquals(double%5B%5D,double%5B%5D,double)
        // public static void assertEquals​(double[] actual, double[] expected, double delta)
        final double delta = 0.001;
        Assert.assertEquals(resultCoordinate.getX(), 674032.357, delta);
        Assert.assertEquals(resultCoordinate.getY(), 6580821.991, delta);
    }
}